service/storage_service.cc

```diff
@@ -1906,48 +1909,174 @@ future<> storage_service::do_stop_ms() {
     });
 }
 
-future<> storage_service::node_ops_cmd_heartbeat_updater(node_ops_cmd cmd, node_ops_id uuid, std::list<gms::inet_address> nodes, lw_shared_ptr<bool> heartbeat_updater_done) {
-    std::string ops;
-    if (cmd == node_ops_cmd::decommission_heartbeat) {
-        ops = "decommission";
-    } else if (cmd == node_ops_cmd::removenode_heartbeat) {
-        ops = "removenode";
-    } else if (cmd == node_ops_cmd::replace_heartbeat) {
-        ops = "replace";
-    } else if (cmd == node_ops_cmd::bootstrap_heartbeat) {
-        ops = "bootstrap";
-    } else {
-        throw std::runtime_error(format("node_ops_cmd_heartbeat_updater: node_ops_cmd is not supported"));
+class node_ops_ctl {
+    std::unordered_set<gms::inet_address> nodes_unknown_verb;
+    std::unordered_set<gms::inet_address> nodes_down;
+    std::unordered_set<gms::inet_address> nodes_failed;
+
+public:
+    const storage_service& ss;
+    sstring desc;
+    locator::host_id host_id;   // Host ID of the node operand (i.e. added, replaced, or leaving node)
+    inet_address endpoint;      // IP address of the node operand (i.e. added, replaced, or leaving node)
+    std::unordered_set<gms::inet_address> sync_nodes;
+    std::unordered_set<gms::inet_address> ignore_nodes;
+    node_ops_cmd_request req;
+    abort_source as;
+    std::optional<future<>> heartbeat_updater_done_fut;
```

Comment:
why are all members public?

Reply:
Ah, I see the control loop is still external.

Reply:
I think this future can be made private, but it's really minor.  The whole class is "private" to the current node ops functions...

Reply:
Yeah, I think it is fine as-is.